analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -29,5 +29,101 @@ public class PublicMethodArgumentsShouldBeCheckedForNull : SymbolicRuleCheck
+
+        static bool IsRelevantMethod(SyntaxNode node) =>
+            node is BaseMethodDeclarationSyntax { } method
+            && MethodDereferencesArguments(method);
+
+        static bool IsRelevantPropertyAccessor(SyntaxNode node) =>
+            node is AccessorDeclarationSyntax accessor
+            && IsPropertyAccessorAccessibleFromOtherAssemblies(accessor.Modifiers);
+
+        static bool IsAccessibleFromOtherAssemblies(SyntaxNode node) =>
+            node.AncestorsAndSelf().OfType<MemberDeclarationSyntax>().FirstOrDefault() is { } containingMember
+            && node.Ancestors().OfType<BaseTypeDeclarationSyntax>().FirstOrDefault() is { } containingType
+            && IsMemberAccessibleFromOtherAssemblies(containingMember.Modifiers(), containingType)
+            && IsTypeAccessibleFromOtherAssemblies(containingType.Modifiers);
+
+        static bool IsMemberAccessibleFromOtherAssemblies(SyntaxTokenList modifiers, BaseTypeDeclarationSyntax containingType) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || (containingType is InterfaceDeclarationSyntax && HasNoDeclaredAccessibilityModifier(modifiers));
+
+        static bool IsPropertyAccessorAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || HasNoDeclaredAccessibilityModifier(modifiers);
+
+        static bool IsTypeAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers);
+
+        static bool IsPublicOrProtectedOrProtectedInternal(SyntaxTokenList modifiers) =>
+            modifiers.AnyOfKind(SyntaxKind.PublicKeyword)
+            || (modifiers.AnyOfKind(SyntaxKind.ProtectedKeyword) && !modifiers.AnyOfKind(SyntaxKind.PrivateKeyword));
+
+        static bool HasNoDeclaredAccessibilityModifier(SyntaxTokenList modifiers) =>
+            !modifiers.Any(x => x.IsAnyKind(
+                SyntaxKind.PrivateKeyword,
+                SyntaxKindEx.FileKeyword,
+                SyntaxKind.ProtectedKeyword,
+                SyntaxKind.InternalKeyword,
+                SyntaxKind.PublicKeyword));
+
+        static bool MethodDereferencesArguments(BaseMethodDeclarationSyntax method)
+        {
+            var argumentNames = method.ParameterList.Parameters
+                                    .Where(x => !x.Modifiers.AnyOfKind(SyntaxKind.OutKeyword))
+                                    .Select(x => x.GetName())
+                                    .ToArray();
+
+            if (!argumentNames.Any())
+            {
+                return false;
+            }
+
+            var walker = new ArgumentDereferenceWalker(argumentNames);
+            walker.SafeVisit(method);
+            return walker.DereferencesMethodArguments;
+        }
+    }
+
+    private sealed class ArgumentDereferenceWalker : SafeCSharpSyntaxWalker
+    {
+        private readonly string[] argumentNames;
```

Comment:
This is not a hashset based on intentional decision that there are typically just a few of them and array tends to be faster for those numbers?

Reply:
I didn't think of it, but I like that explanation 😁